File: 백준/Gold/16168. 퍼레이드/퍼레이드.cc
#include <iostream>
#include <vector>
#include <set>
using namespace std;

vector<vector<int>> v;
int vis[3001][3001];
int n, m;

void go(int node, int idx, int cnt)
{
	if (cnt == m)
	{
		cout << "YES";
		exit(0);
	}
	for (auto &i : v[node])
	{
		if (vis[i][node] != idx && vis[node][i] != idx)
		{
			vis[i][node] = vis[node][i] = idx;
			go(i, idx, cnt + 1);
		}
	}
}

int main()
{
	cin >> n >> m;
	v.resize(n + 1);
	for (int i = 0; i < m; i++)
	{
		int a, b;
		cin >> a >> b;
		v[a].push_back(b);
		v[b].push_back(a);
	}

	for (int i = 1; i <= n; i++)
	{
		go(i, i, 0);
	}

	cout << "NO";
}

File: 백준/Gold/16168. 퍼레이드/퍼레이드.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.List;
import java.util.StringTokenizer;

public class Main {

	static int n, m;
	static List<List<Integer>> v = new ArrayList<>();
	static int[][] e;
	static String ans = "NO";

	public static void main(String[] args) throws Exception {
		var br = new BufferedReader(new InputStreamReader(System.in));

		var st = new StringTokenizer(br.readLine());
		n = Integer.parseInt(st.nextToken());
		m = Integer.parseInt(st.nextToken());

		for (int i = 0; i <= n; i++) {
			v.add(new ArrayList<>());
		}
		for (int i = 0; i < m; i++) {
			st = new StringTokenizer(br.readLine());
			int a = Integer.parseInt(st.nextToken());
			int b = Integer.parseInt(st.nextToken());
			v.get(a).add(b);
			v.get(b).add(a);
		}

		e = new int[n + 1][n + 1];
		for (int i = 1; i <= n; i++) {
			go(i, 0, 0, i);
		}

		System.out.println(ans);
	}

	static void go(int node, int pre, int cnt, int ver) {
		if (cnt == m) {
			ans = "YES";
			return;
		}
		for (var i : v.get(node)) {
			if (e[node][i] == ver || e[i][node] == ver || i == pre) {
				continue;
			}
			e[node][i] = e[i][node] = ver;
			go(i, node, cnt + 1, ver);
		}
	}
}